jax/interpreters/xla.py

```diff
@@ -233,17 +233,38 @@ def device_persistent_result_handler(result_shape):
   else:
     raise TypeError(t)
 
-def pyval_result_handler(result_shape):
+def _pyval_result_prefetcher(result_shape):
+  t = type(result_shape)
+  if t is ResultArray:
+    return lambda buf: _copy_to_host_async(buf)
+  elif t is ResultTuple:
+    _, result_shapes = result_shape
+    handlers = list(map(_pyval_result_prefetcher, result_shapes))
+    def f(buf):
+      for h, b in zip(handlers, buf.destructure()):
```

Comment:
Mainly for my understanding: could we instead just execute a prefetch on the tuple itself (and not destructure here)? In that case, we might not need separate prefetcher functions based on the result shape (i.e. tuple vs array) and we could simplify this logic.

Reply:
We can. Done.